File: bindings/node/src/zlink/contracts/service/spot/spot_route_bridge.ts
// SPDX-License-Identifier: MPL-2.0

import type { RoutingId } from '../../core';
import type { MessageLike } from '../../messaging';
import type { RouterSocket } from '../../sockets';
import type { RequestOperation, SendOperation } from './spot_operations';

export const SpotRouteBridgeEndpointCapabilities = {
  SpotRoute: 0x00000001,
  RouteOnly: 0x00000001
} as const;

export type SpotRouteBridgeEndpointCapabilitiesValue =
  typeof SpotRouteBridgeEndpointCapabilities[keyof typeof SpotRouteBridgeEndpointCapabilities];

export interface SpotRouteBridgeEndpointOptions {
  readonly capabilities?: SpotRouteBridgeEndpointCapabilitiesValue | number;
}

export interface SpotRouteBridge {
  attachRouterChannel(channelName: string, router: RouterSocket, options?: SpotRouteBridgeEndpointOptions): void;
  send(channelName: string, targetNodeRid: RoutingId, targetSpotRid: RoutingId): SendOperation;
  request(channelName: string, targetNodeRid: RoutingId, targetSpotRid: RoutingId): RequestOperation;
  handleRouterReceived(
    channelName: string,
    sourceNodeRid: RoutingId,
    requestSeq: bigint | number,
    parts: MessageLike | readonly MessageLike[]
  ): boolean;
  close(): void;
}

export interface SpotNodePublisher {
  publish(topic: string): SendOperation;
  close(): void;
}

File: bindings/java/src/main/java/systems/zlink/contracts/service/spot/SpotRouteBridge.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.service.spot;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.contracts.sockets.RouterSocket;
import java.util.List;

/** Bridges caller-owned channel sockets to the local SPOT routed plane. */
public interface SpotRouteBridge extends AutoCloseable {
    void attachRouterChannel(String channelName, RouterSocket router);

    void attachRouterChannel(
        String channelName,
        RouterSocket router,
        SpotRouteBridgeEndpointOptions options);

    SendOperation send(
        String channelName,
        RoutingId targetNodeRid,
        RoutingId targetSpotRid);

    RequestOperation request(
        String channelName,
        RoutingId targetNodeRid,
        RoutingId targetSpotRid);

    boolean handleRouterReceived(
        String channelName,
        RoutingId sourceNodeRid,
        long requestSeq,
        List<Message> parts);

    int drain();

    void close();
}
